functionary/vllm_monkey_patch/async_llm_engine.py

```diff
@@ -0,0 +1,724 @@
+        """
+        # Prompt template v2 directly begins with "function" stage while
+        # Prompt template v1 needs to generate <|START_OF_FUNCTION_CALL|> first
+        # Future versions are assumed to begin directly with "function" stage
+        self.gen_states[request_id] = {
+            "stage": "function"
+            if self.prompt_templates[request_id].version not in ["v1"]
+            else "pre-function",
+            "curr_tokens": [],
+            "curr_text": "",
+            "func_name": "",
+            "param_names": [],
+        }
+
+    def update_gen_state(self, request_id, new_token_id):
+        """This function updates the generation state by calling the respective
+        prompt_template.update_grammar_sampling_gen_state class. Each prompt
+        template has its own way of parsing the generation to figure out the
+        current state of generation.
+        """
+        gen_state = self.gen_states[request_id]
+        prompt_template = self.prompt_templates[request_id]
+
+        # Form the functions/parameters options
+        if gen_state["stage"] in ["pre-function", "function"]:
+            options = [
+                tool_or_func["name"]
+                for tool_or_func in self.tools_or_functions[request_id]
+            ]
+        else:
+            func_name = gen_state["func_name"]
+            for tool_or_func in self.tools_or_functions[request_id]:
+                if tool_or_func["name"] == func_name:
+                    options = list(tool_or_func["parameters"]["properties"].keys())
+                    break
+
+        # Update the gen_state
+        self.gen_states[request_id] = prompt_template.update_grammar_sampling_gen_state(
+            gen_state=gen_state,
+            new_token_id=new_token_id,
+            options=options,
+            tokenizer=self.tokenizer,
+        )
+
+    def sample(
+        self,
+        request_id,
+        delta_token_ids,
+        model_sampled_token_id,
+    ):
+        """Applies grammar-sampling to the token generation and returns a
+        newly sampled token.
+
+        This function checks whether the model-sampled token helps towards
+        forming one of the function names or parameter names. It loops through
+        a list of token ids sorted in descending order by the log probabilities.
+        It replaces the output token if the grammar-sampled token is different
+        from the model-sampled token
+        """
+        gen_state = self.gen_states[request_id]
+        prompt_template = self.prompt_templates[request_id]
+
+        # Only perform grammar sampling if the gen_state is one of ["function", "parameter-name"]
+        if gen_state["stage"] == "function":
+            # Get the list of functions
+            options = [
+                tool_or_func["name"]
+                for tool_or_func in self.tools_or_functions[request_id]
+            ]
+            if self.prompt_templates[request_id].version not in ["v1"]:
```

Comment:
Can we add another method in prompt template for getting list of additional predefined function name with default implementation is return []

def get_predefined_function_names() --> List[str]
in base_template:
def get_predefined_function_names():
   return []
in template_v2:
def get_predefined_function_names()
   return ["all"]

Reply:
Gotcha